openfl/callbacks/callback_list.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2020-2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+from typing import List
+
+from openfl.callbacks.callback import Callback
+from openfl.callbacks.memory_profiler import MemoryProfiler
+
+
+class CallbackList(Callback):
+    """An ensemble of callbacks.
+
+    This class allows multiple callbacks to be used together, by sequentially
+    calling each callback's respective methods.
+
+    Attributes:
+        callbacks: A list of `openfl.callbacks.Callback` instances.
+        add_memory_profiler: If True, adds a `MemoryProfiler` callback to the list.
+        tensor_db: Optional `TensorDB` instance of the respective participant.
+            If provided, callbacks can access TensorDB for various actions.
+        params: Additional parameters saved for use within the callbacks.
+    """
+
+    def __init__(
+        self,
+        callbacks: List[Callback],
+        add_memory_profiler=False,
+        tensor_db=None,
+        **params,
+    ):
+        super().__init__()
+        self.callbacks = _flatten(callbacks) if callbacks else []
```

Comment:
Could you give an example/explain how this flatten would be used given that `callbacks` arg has list typing?

Reply:
It is to cover an edge case, where `callbacks` is a single `Callback` instead of a sequence of `Callback`s. `_flatten` will then create a singleton list.

I forgot to add that, it is now added in https://github.com/securefederatedai/openfl/pull/1195/commits/af0c40fe9e2a488daeddb4817b07299d8ec0f5d7